contrib/opencensus-ext-azure/opencensus/ext/azure/common/__init__.py

```diff
@@ -17,14 +17,79 @@
 
 from opencensus.ext.azure.common.protocol import BaseObject
 
+INGESTION_ENDPOINT = 'IngestionEndpoint'
+INSTRUMENTATION_KEY = 'InstrumentationKey'
+
+
+def process_options(options):
+    code_cs = parse_connection_string(options.connection_string)
+    code_ikey = options.instrumentation_key
+    env_cs = parse_connection_string(
+        os.getenv('APPLICATIONINSIGHTS_CONNECTION_STRING'))
+    env_ikey = os.getenv('APPINSIGHTS_INSTRUMENTATIONKEY')
+
+    # The priority of which value takes on the instrumentation key is:
+    # 1. Key from explicitly passed in connection string
+    # 2. Key from explicitly passed in instrumentation key
+    # 3. Key from connection string in environment variable
+    # 4. Key from instrumentation key in environment variable
+    options.instrumentation_key = code_cs.get(INSTRUMENTATION_KEY) \
+        or code_ikey \
+        or env_cs.get(INSTRUMENTATION_KEY) \
+        or env_ikey
+    # The priority of the ingestion endpoint is as follows:
+    # 1. The endpoint explicitly passed in connection string
+    # 2. The endpoint from the connection string in environment variable
+    # 3. The default breeze endpoint
+    endpoint = code_cs.get(INGESTION_ENDPOINT) \
+        or env_cs.get(INGESTION_ENDPOINT) \
+        or 'https://dc.services.visualstudio.com'
+    options.endpoint = endpoint + '/v2/track'
+
+
+def parse_connection_string(connection_string):
+    if connection_string is None:
+        return {}
+    try:
+        pairs = connection_string.split(';')
+        result = dict(s.split('=') for s in pairs)
+    except Exception:
+        raise ValueError('Invalid connection string')
+    # Validate authorization
+    auth = result.get('Authorization')
+    if auth is not None and auth.lower() != 'ikey':
+        raise ValueError('Invalid authorization mechanism')
+    # Construct the ingestion endpoint if not passed in explicitly
+    if result.get(INGESTION_ENDPOINT) is None:
+        endpoint_suffix = ''
+        location_prefix = ''
+        suffix = result.get('EndpointSuffix')
```

Comment:
I think the keys are supposed to be case-insensitive. Probably not a blocker since I don't see this actually being tested in other SDKs, but it could lead to some confusing errors for users that manually modify their conn strs :thinking:

Reply:
Good point. It'll be safer to be case-insensitive.